addons/mrp/report/mrp_report_bom_structure.py

```diff
@@ -560,13 +560,20 @@ def _get_resupply_route_info(self, warehouse, product, quantity, bom=False):
         found_rules = []
         if self._need_special_rules(self.env.context.get('product_info'), self.env.context.get('parent_bom'), self.env.context.get('parent_product_id')):
             found_rules = self._find_special_rules(product, self.env.context.get('product_info'), self.env.context.get('parent_bom'), self.env.context.get('parent_product_id'))
+            if found_rules and not self._is_resupply_rules(found_rules, bom):
```

Comment:
Not sure to understand why we test the rules list with `any` instead of filter out non-wanted rules ?
Also, if it's about subcontracting, shouldn't it more efficient to fix `_find_special_rules` directly instead ?

Reply:
Initially, it was because it needs to check if a `buy` route exists, which isn't possible by itself with only `mrp_subcontracting`. But nothing stops me from directly calling that check in `_find_secial_rules()` indeed. Would make the code simpler to read, thanks for the suggestion!

As to not filter out the unwanted rules, since we fetch the rules with `_get_rules_from_location()` (but here on the subcontracting location), we simply don't have that option. :slightly_frowning_face: